### PR TITLE
Feature/boolean slider

### PR DESCRIPTION
**Pull request by edwinmp**

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                "./dist/MxTestProject/deployment/web/widgets/" + pkg.name + "/*",
+                "./dist/MxTestProject/widgets/" + pkg.name + ".mpk"
+            ]
+        }
+    });
+
+    grunt.loadNpmTasks("grunt-contrib-compress");
+    grunt.loadNpmTasks("grunt-contrib-clean");
+    grunt.loadNpmTasks("grunt-contrib-watch");
+    grunt.loadNpmTasks("grunt-contrib-copy");
```
Comment (Andries-Smit) [inline Gruntfile.js:80]: Add dependency check

File: karma.conf.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        "react/lib/ExecutionEnvironment",
+        "react/lib/ReactContext",
+        "react/addons",
+        "jsdom"
+    ])
+});
+
+module.exports = function(config) {
+    if (config.codeCoverage) {
+        console.log("With instrumenter");
```
Comment (Andries-Smit) [inline karma.conf.js]: Remove console.log

File: package.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
+{
+  "name": "booleanslider",
+  "version": "0.1.0",
+  "description": "Toggle a boolean attribute.",
```
Comment (Andries-Smit) [inline package.json]: Remove period.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  },
+  "devDependencies": {
+    "@types/big.js": "0.0.31",
+    "@types/classnames": "0.0.32",
+    "@types/core-js": "^0.9.35",
+    "@types/dojo": "^1.9.37",
+    "@types/enzyme": "^2.7.2",
+    "@types/jasmine": "^2.5.41",
+    "@types/karma": "^0.13.33",
+    "@types/react": "^15.0.4",
```
Comment (Andries-Smit) [inline package.json]: Update typing

File: src/com/mendix/widget/BooleanSlider/BooleanSlider.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    uninitialize(): boolean {
+        unmountComponentAtNode(this.domNode);
+
+        return true;
+    }
+
+    private updateRendering(alertMessage?: string) {
+        const status: SliderStatus = this.contextObject
+            ? !this.readOnly && !this.contextObject.isReadonlyAttr(this.booleanAttribute) ? "enabled" : "disabled"
```
Comment (Andries-Smit) [inline src/com/mendix/widget/BooleanSlider/BooleanSlider.ts]: Extract for readability
`const enabled = !this.readOnly && !this.contextObject.isReadonlyAttr(this.booleanAttribute);`

File: src/com/mendix/widget/BooleanSlider/BooleanSlider.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    private handleToggle() {
+        this.contextObject.set(this.booleanAttribute, !this.contextObject.get(this.booleanAttribute));
+        this.executeAction(this.onChangeMicroflow, this.contextObject.getGuid());
+    }
+
+    private executeAction(actionname: string, guid: string) {
+        if (actionname) {
+            window.mx.ui.action(actionname, {
+                error: (error: Error) =>
```
Comment (Andries-Smit) [inline src/com/mendix/widget/BooleanSlider/BooleanSlider.ts]: Remove type `Error` it is defined in mendix-client typing `mx.ui.action`

File: src/com/mendix/widget/BooleanSlider/components/Slider.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            type: "checkbox"
+        }),
+        DOM.div({
+            className: classNames("widget-boolean-slider-btn", {
+                "enabled": props.status === "enabled",
+                "no-slider": props.status === "no-context"
+            }),
+            onClick: props.status === "enabled" ? props.onClick : undefined
+        }),
+        props.alertMessage ? createElement(Alert, { message: props.alertMessage }) : null
```
Comment (Andries-Smit) [inline src/com/mendix/widget/BooleanSlider/components/Slider.ts]: Move condition to Alert component

File: src/com/mendix/widget/BooleanSlider/components/Alert.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+import { DOM } from "react";
+
+export const Alert = (props: { message: string }) =>
+    DOM.div({ className: "alert alert-danger widget-boolean-slider-alert" }, props.message);
```
Comment (Andries-Smit) [inline src/com/mendix/widget/BooleanSlider/components/Alert.ts]: Add condition
```
export const Alert = (props: { message: string }) =>
    props.message 
        ? DOM.div({ className: "alert alert-danger widget-boolean-slider-alert" }, props.message)
        : null
```

File: src/com/mendix/widget/BooleanSlider/components/__tests__/Alert.spec.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+describe("Alert", () => {
+    const message = "This is an error";
+    const alert = shallow(createElement(Alert, { message }));
+
+    it("renders structure correctly", () => {
+        expect(alert).toBeElement(
+            DOM.div({ className: "alert alert-danger widget-boolean-slider-alert" }, message)
+        );
+    });
```
Comment (Andries-Smit) [inline src/com/mendix/widget/BooleanSlider/components/__tests__/Alert.spec.ts]: Add test for condition no prop.message

File: src/com/mendix/widget/BooleanSlider/components/Slider.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        DOM.input({
+            checked: props.isChecked,
+            className: classNames("widget-boolean-slider-checkbox", { enabled: props.status === "enabled" }),
+            readOnly: true,
+            type: "checkbox"
+        }),
+        DOM.div({
+            className: classNames("widget-boolean-slider-btn", {
+                "enabled": props.status === "enabled",
+                "no-slider": props.status === "no-context"
```
Comment (Andries-Smit) [inline src/com/mendix/widget/BooleanSlider/components/Slider.ts:28]: Is this the correct name `no-slider`?

**Review by Andries-Smit — CHANGES_REQUESTED** [review]
Good quality code, last small steps before approval

File: src/com/mendix/widget/BooleanSlider/ui/Slider.sass [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            height: 100%
+
+
+        &:after
+            left: 0
+            border-radius: 2em
+            background: #fbfbfb
+            transition: left .3s cubic-bezier(0.175, 0.885, 0.320, 1.275), padding .3s ease, margin .3s ease
+            box-shadow: 0 0 0 1px rgba(0, 0, 0, .1), 0 4px 0 rgba(0, 0, 0, .08)
+
```
Comment (Andries-Smit) [inline src/com/mendix/widget/BooleanSlider/ui/Slider.sass:47]: Remove double white lines and spaces in the sass files

File: package.json (hunk 3 above)
Comment (edwinmp) [reply package.json]: Done.

File: src/com/mendix/widget/BooleanSlider/BooleanSlider.ts (hunk 5 above)
Comment (edwinmp) [reply src/com/mendix/widget/BooleanSlider/BooleanSlider.ts]: Done.

File: src/com/mendix/widget/BooleanSlider/BooleanSlider.ts (hunk 6 above)
Comment (edwinmp) [reply src/com/mendix/widget/BooleanSlider/BooleanSlider.ts]: Done.

File: src/com/mendix/widget/BooleanSlider/components/Slider.ts (hunk 7 above)
Comment (edwinmp) [reply src/com/mendix/widget/BooleanSlider/components/Slider.ts]: Done.

File: src/com/mendix/widget/BooleanSlider/components/Alert.ts (hunk 8 above)
Comment (edwinmp) [reply src/com/mendix/widget/BooleanSlider/components/Alert.ts]: Done.

File: src/com/mendix/widget/BooleanSlider/components/__tests__/Alert.spec.ts (hunk 9 above)
Comment (edwinmp) [reply src/com/mendix/widget/BooleanSlider/components/__tests__/Alert.spec.ts]: Done.

File: src/com/mendix/widget/BooleanSlider/ui/Slider.sass (hunk 11 above)
Comment (edwinmp) [reply src/com/mendix/widget/BooleanSlider/ui/Slider.sass:47]: Done.

File: karma.conf.js (hunk 2 above)
Comment (edwinmp) [reply karma.conf.js]: Done.

File: Gruntfile.js (hunk 1 above)
Comment (edwinmp) [reply Gruntfile.js:80]: Not sure what you meant here :)

File: src/com/mendix/widget/BooleanSlider/components/Slider.ts (hunk 10 above)
Comment (edwinmp) [reply src/com/mendix/widget/BooleanSlider/components/Slider.ts:28]: Yup... A better suggestion?

**Pull request closed without merging**